server/keyshare/email.go

```diff
@@ -150,21 +150,33 @@ func VerifyMXRecord(email string) error {
 
 	host := email[strings.LastIndex(email, "@")+1:]
 
-	if records, err := net.LookupMX(host); err != nil || len(records) == 0 {
-		if err != nil {
-			if derr, ok := err.(*net.DNSError); ok && (derr.IsTemporary || derr.IsTimeout) {
-				// When DNS is not resolving or there is no active network connection
-				server.Logger.WithField("error", err).Error("No active network connection")
-				return ErrNoNetwork
-			}
+	records, err := net.LookupMX(host)
+
+	if err != nil || len(records) == 0 {
+		if derr, ok := err.(*net.DNSError); ok && (derr.IsTemporary || derr.IsTimeout) {
+			// When DNS is not resolving or there is no active network connection
+			server.Logger.WithField("error", err).Error("No active network connection")
+			return ErrNoNetwork
 		}
 
 		// Check if there is a valid A or AAAA record which is used as fallback by mailservers
 		// when there are no MX records present
 		if records, err := net.LookupIP(host); err != nil || len(records) == 0 {
 			return ErrInvalidEmailDomain
 		}
-		return nil
 	}
+
+	invalidHosts := 0
+	for _, h := range records {
+		// Check if host specified at MX record valid
+		if addr, err := net.LookupHost(h.Host); err != nil || len(addr) == 0 {
+			invalidHosts++
+		}
+	}
+
+	if invalidHosts >= len(records) {
+		return ErrInvalidEmailDomain
+	}
```

Comment:
You assumed here that the mail relay does not return an error if at least one of the hosts in the MX has a valid value? Do we know this for sure? Otherwise, it could be much simpler and we could already do an early return if at least one of the hosts is invalid.